AirCasting/SessionsSynchronization/Adapters/Services/SessionDownloadService.swift

```diff
@@ -0,0 +1,68 @@
+// Created by Lunar on 10/06/2021.
+//
+
+import Foundation
+import Combine
+
+final class SessionDownloadService: SessionDownstream {
+    private let client: APIClient
+    private let authorization: RequestAuthorisationService
+    private let responseValidator: HTTPResponseValidator
+    
+    private let decoder: JSONDecoder = {
+        let decoder = JSONDecoder()
+        decoder.keyDecodingStrategy = .convertFromSnakeCase
+        let formatter = DateFormatter()
+        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"
+        decoder.dateDecodingStrategy = .formatted(formatter)
+        return decoder
+    }()
+    
+    init(client: APIClient, authorization: RequestAuthorisationService, responseValidator: HTTPResponseValidator) {
+        self.client = client
+        self.authorization = authorization
+        self.responseValidator = responseValidator
+    }
+    
+    func download(session: SessionUUID) -> AnyPublisher<SessionsSynchronization.SessionDownstreamData, Error> {
```

Comment:
Are You sure that this is possible and easy to test? Imo completion block here and a wrapper on publisher is more reelable/redable/testable

Reply:
I'm quite sure since it indeed is tested. Completion block would mess the code inside `SessionSynchronizationController` pretty bad too, because basically the purpose of this class is to adjust existing API to the needs of `SessionSynchronizationController` which has Combine all over. We could introduce a new class that would communicate with the API and have completion-block interface and here only adjust for Combine, but I think splitting it into a private method here will have the same effect (will be readable on the same level) and won't clutter the project. I'll also change the Upstream implementation too since it uses the same pattern